# The Mathematics of Scorigami


## Introduction:
On Monday Night Football, September 25, the Philadelphia Eagles defeated the Tampa Bay Buccaneers by a score of 25-11. This event marked the most recent [Scorigami](https://nflscorigami.com/): a game score which had not previously been recorded in the NFL. Scorigamis are rare but have become one of the most entertaining outcomes to cheer for in a game. To see an event that has never happened before is a remarkable feature, especially given that 17,428 NFL games have been played up to this point. 

I began pondering the probability of Scorigami's occuring, and realized that exploring this question would touch on some very important and interesting mathematical concepts. What was the probability of that Monday Night Football game ending in a score of 25-11? Which Scorigamis that we have not yet observed are most likely to occur next? 

To begin with, let's start with some seemingly trivial notions. First, a score is a tuple $(a, b)$ where $a$ is the number of points scored by team 1 and $b$ is the number of points scored by team 2. The number of points scored by each team is a nonnegative integer. Given that the highest NFL score in a game ever was 73 points in 1940 (a 73-0 victory by the Chicago Bears over the Washington Redskins), we can very safely assume that each score lies in the interval $[0, 100]$. 

Now, we want to do something seemingly simple: we want to understand what the probability that a given score occurs is. From this point on, there is something deceptive at work here. In the first way of interpreting this situation, team 1 scores $a$ points and team 2 scores $b$ points. The team with more points is declared the winner (or it is a tie) and the score is recorded as $(a, b)$. 

In the second way of interpreting this situation, team 1 plays team 2, and the outcome of the game is a score $(a, b)$. 

Why are these different? And why does it matter when it comes to Scorigami? 

In the first situation, we imagine each team scoring a certain number of points--$\textit{independent}$ of what the other team scored. So, we can consider the probability of a score $(a, b)$ to happen as the probability that team 1 scores $a$ points and the probability that team 2 scores $b$ points. 

In the second interpretation, we are considering the $\textit{joint}$ probability of a given score $(a,b)$ occuring. In this interpretation, fundamental to the score is that $a$ and $b$ may be related. We cannot consider them happening one at a time--the game is played between both, and these are inextricably linked. 

Many games happen seemingly like the first situation--the teams go out and each try to score on every possession. Often, a team that is good at scoring will put up many points, irrespective of what the other teams do. 

But it is clear that this is a simplification of reality. For instance, the amount of scoring you are able to do is certainly dependent on how much you have the ball--this is very different if you opponent has the ball for long, scoring drives, or if they are consistenly punting on 3 and outs. Furthermore, if the regulation ends in a tie, it is unlikely that the score will remain a tie after the game. Finally, there are crazy possibilities in the rules that have never occured like a safety on an extra point that would result in a score of 6-1, meaning that the only possible way to score one point is if the opponent also scores a touchdown. 

Because of these considerations, it is clear that the best way to understand the probability that various scores occur needs to understand the joint probability of each score. However, as we proceed, it will be clear that this presents some major challenges to overcome. 

## Interpretation 1
We will start by examining interpretation 1--namely, that each team scores points independently of the other team. Recall that each team scores a score between $[0, 100]$. We refer to these scores as the outcomes. When we say that the probability that team 1 scores 10 points is 8%, what we really mean is that there is some function $\mathbf{P}$ such that $\mathbf{P}(a = 10) = 0.08$. The probability function $\mathbf{P}$ just assigns a number between 0 and 1 to the outcome that we were interested in--namely, that $a = 10$, or in english, that team 1 scored 10 points. I like to read the notation $\mathbf{P}(a = 10)$ by saying, "The probability that a is 10". There are just a few conditions on the probability functions that we care about--for instance, if we want to calculate the probability that team 2 scores greater than 10 and less than 14 points, we would write it that $$\mathbf{P}(10 < b < 14) = \mathbf{P}(b = 11) + \mathbf{P}(b = 12) + \mathbf{P}(b = 13)$$

## Empirical distribution

Say we want to know what the probability that the Scorigami score of 25-11 occurring is. According to this simplified scenario, this would be equal to the probability that team 1 scores 25 points and team 2 scores 11 points (or team 1 scores 11, team 2 scores 25). Since these are independent events, we can just multiply them: 
$$\mathbf{P}((a, b) = (25, 11)) = \mathbf{P}(a = 25)* \mathbf{P}(b = 11)$$
However, how do we know what the probability that a team scores 25 points is? It's not a very common score. One way of doing this is by constructing what is known as an empirical probability distribution. Empirical comes from the fact that we look at events that have already happened--NFL games--to try to understand how likely scoring is. We are assuming that there is some underlying law $\mathbf{P}$ that says how likely scoring any number of points is, and to get a good guess at what $\mathbf{P}$ is, we construct our empirical probability distribution, which to distinguish, I will denote $\mathbf{P}_e$. 

The empirical probability distribution is defined in the most natural way we can think of--we assume that the probability that a certain score occurs is the frequency of it having occured in the past. Since there are 17,428 NFL games ever, there are 34,856 individual team scores that have been recorded. That's a lot! If we want to estimate the probability that a team scores 10 points in a game, we simply take the number of times a score of 10 has been observed, 1,952, and divide it by 34,856 to get 5.60%. We would write this as 
$$\mathbf{P}_e(a = 10) = \frac{1952}{34856} \approx 0.0560$$

To find our empirical probability distribution, we're gonna use the help of our computer. 

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [2]:
df = pd.read_csv("NFLGames.csv")

In [3]:
df.head()

Rk  Score  PtsW  PtsL  PtTot  PD  Count Unnamed: 7  \
0   1  20-17    20    17     37   3    284  all games   
1   2  27-24    27    24     51   3    232  all games   
2   3  17-14    17    14     31   3    200  all games   
3   4  23-20    23    20     43   3    199  all games   
4   5  24-17    24    17     41   7    174  all games   

                                           Last Game  
0  New Orleans Saints vs. Carolina Panthers Septe...  
1  Tennessee Titans vs. Los Angeles Chargers Sept...  
2  Los Angeles Chargers vs. Tennessee Titans Dece...  
3  Kansas City Chiefs vs. Cincinnati Bengals Janu...  
4  Miami Dolphins vs. New England Patriots Septem...

In [4]:
df = df.drop(columns = ['Rk', 'PtTot', 'PD', 'Unnamed: 7', 'Last Game'])

In [5]:
highest_score = max(df['PtsW'])
highest_score

73

In [6]:
'''Here, we construct an array called counts, 
so that index i of the array is the number of times 
score i has been recorded--from 0 to 73''' 

counts = [0]*(highest_score+1)

for index, row in df.iterrows():
    counts[row['PtsW']] += row['Count']
    counts[row['PtsL']] += row['Count']

In [7]:
print(counts)
num_scores = sum(counts)
num_games = num_scores/2
num_games
num_scores

[1534, 0, 37, 935, 1, 21, 877, 1873, 47, 487, 1952, 57, 326, 1630, 1993, 242, 1087, 2471, 153, 636, 2089, 1636, 368, 1241, 2105, 224, 614, 1658, 1144, 274, 816, 1297, 143, 366, 905, 579, 146, 422, 638, 60, 144, 360, 268, 72, 134, 251, 21, 54, 119, 89, 17, 50, 51, 9, 12, 27, 33, 7, 8, 14, 2, 5, 9, 3, 1, 2, 2, 0, 0, 0, 2, 0, 1, 1]


34852

In [8]:
# Now, we write a function to empirically compute the probability that team 1 scores a points and team 2 scores b points
def score_probability(a, b, distribution = counts, total = num_scores):
    return(distribution[a]*distribution[b]/(total**2))

# We have another function which gives us the probability that the final score is (a, b), without caring which team scored which
def score_P(a, b):
    if a == b:
        return score_probability(a, b)
    return 2*score_probability(a, b)
MNF_game = score_P(25, 11) 

print("Probability of a final score of 25-11: {:.4%}".format(MNF_game))

Probability of a final score of 25-11: 0.0021%


Wow, that is small! One fun probability fact is that if you have a sequence of events $x_1, x_2, ...$ which all have the same probability distribution, and the probability of an event $a$ you care about occuring is $p$, then you are expected to wait $1/p$ times until the event $a$ occurs. So, with the probability that we calculated, we get: 

In [9]:
1/MNF_game

47566.64724310777

Wow! We expect to wait about 47567 games to see a score of 25-11 under our empirical probability distribution. That means that this score happened a lot sooner than we would have expected!

Next, let's compare what our empirical distribution predicts about each of the observed scores. How many times would we have expected to see the scores that were observed? 

In [10]:
df['Score Probability'] = df.apply(lambda row: score_P(row['PtsW'], row['PtsL']), axis=1)
df['Expected Occurrences'] = df.apply(lambda row: row['Score Probability']*num_games, axis = 1)
f"There are {df['Expected Occurrences'].sum():.1f} total games expected to be played with existing scores"

'There are 17127.9 total games expected to be played with existing scores'

We have played 17,428 games ever, yet only 17,128 appear in the empirical distribution of the existing scores. What does this mean? Our empirical distribution predicts that a whopping 300 games played up to this point should have a score which has never occurred.

In [11]:
df['Empirical error'] = df.apply(lambda row: abs(row['Count'] - row['Expected Occurrences']), axis = 1)
df.head(20)

Score  PtsW  PtsL  Count  Score Probability  Expected Occurrences  \
0   20-17    20    17    284           0.008499            148.109692   
1   27-24    27    24    232           0.005747            100.140308   
2   17-14    17    14    200           0.008109            141.303311   
3   23-20    23    20    199           0.004269             74.384512   
4   24-17    24    17    174           0.008564            149.244089   
5   13-10    13    10    167           0.005239             91.293470   
6   24-21    24    21    156           0.005670             98.811546   
7   17-10    17    10    144           0.007942            138.396419   
8   16-13    16    13    143           0.002917             50.838115   
9   24-14    24    14    139           0.006908            120.373723   
10  24-10    24    10    135           0.006766            117.897395   
11  23-17    23    17    126           0.005049             87.986658   
12  27-20    27    20    125           0.005703             99.379146   
13  27-17    27    17    122           0.006746            117.551877   
14  20-10    20    10    120           0.006714            117.001262   
15  24-20    24    20    117           0.007240            126.171956   
16  20-13    20    13    113           0.005607             97.700849   
17   17-7    17     7    108           0.007621            132.795335   
18  17-13    17    13    106           0.006632            115.566682   
19  21-17    21    17    104           0.006656            115.992081   

    Empirical error  
0        135.890308  
1        131.859692  
2         58.696689  
3        124.615488  
4         24.755911  
5         75.706530  
6         57.188454  
7          5.603581  
8         92.161885  
9         18.626277  
10        17.102605  
11        38.013342  
12        25.620854  
13         4.448123  
14         2.998738  
15         9.171956  
16        15.299151  
17        24.795335  
18         9.566682  
19        11.992081

Just looking at the data above makes it clear that this empirical probability distribution--namely, the model of considering scores happening independently--is obviously far from correct. It vastly underestimated the most common score of 20-17. If we look at the tail--the least common games--there are a few game scores which are predicted to occur much more often than they have.

In [12]:
df.tail(20)

Score  PtsW  PtsL  Count  Score Probability  Expected Occurrences  \
1057  55-17    55    17      1       1.098528e-04              1.914295   
1058   44-3    44     3      1       2.062961e-04              3.594916   
1059  44-15    44    15      1       5.339428e-05              0.930449   
1060   13-8    13     8      1       1.261421e-04              2.198152   
1061  36-32    36    32      1       3.437664e-05              0.599047   
1062  44-33    44    33      1       8.075334e-05              1.407208   
1063  51-13    51    13      1       1.341937e-04              2.338460   
1064  48-46    48    46      1       4.114725e-06              0.071703   
1065   36-9    36     9      1       1.170729e-04              2.040112   
1066   10-5    10     5      1       6.749533e-05              1.176174   
1067   19-5    19     5      1       2.199130e-05              0.383220   
1068  48-40    48    40      1       2.821526e-05              0.491679   
1069  55-40    55    40      1       6.401781e-06              0.111557   
1070  46-23    46    23      1       4.291071e-05              0.747762   
1071    8-6     8     6      1       6.786909e-05              1.182687   
1072    7-5     7     5      1       6.476370e-05              1.128572   
1073  59-21    59    21      1       3.771255e-05              0.657179   
1074  72-41    72    41      1       5.927575e-07              0.010329   
1075   33-9    33     9      1       2.934841e-04              5.114255   
1076  38-12    38    12      1       3.424624e-04              5.967749   

      Empirical error  
1057         0.914295  
1058         2.594916  
1059         0.069551  
1060         1.198152  
1061         0.400953  
1062         0.407208  
1063         1.338460  
1064         0.928297  
1065         1.040112  
1066         0.176174  
1067         0.616780  
1068         0.508321  
1069         0.888443  
1070         0.252238  
1071         0.182687  
1072         0.128572  
1073         0.342821  
1074         0.989671  
1075         4.114255  
1076         4.967749

Are these just statistical abberations? Random chance? How bad are these discrepancies, really? Let's take the most common game score, 20-17, as an example. The empirical model predicts that this score occurs with a probability of 0.8499%, so that in all games that have been played, we would expect to see it 148 times--but it's happened 284 times. If the true probability of a 20-17 score was 0.8499%, what would the probability be of having it happen at least 284 times in 17,428 games? This is a classical statistics problem known as finding the p-value. If an event is repeated $n$ times, and the probability of occurring at any one trial is $p$, then the probability of $k$ successes is given according to the Binomial distribution: 
$$\mathbf{P}(x = k) = {n \choose k} (p)^k(1-p)^{n-k}$$
This takes a while to calculate if we need to add up the probabilities for $x = 284, 285, ..., 17428$ (or just adding up $x = 0, 1, ..., 283$ and subtracting this from 1). Fortunately, the binomial distribution approaches the normal distribution (the bell curve shape) as $n$ gets big enough. As a rule of thumb, to apply this approximation, you want $np \geq 5$ and $n(1-p)\geq 5$. Since $np = 17428*(0.008499) = 148.1$, we are safe to use this approximation. We want to find the probability that a Normal random variable with mean $np$ and a standard deviation equal to $\sqrt{np(1-p)}$ has a value at least as large as 284 (since this is a continuous variable, we will take its value to be at least 283.5). 

In [13]:
import scipy.stats as stats
import math

n = num_games
p = df['Score Probability'][0]

mean = n*p
std_dev = math.sqrt(n*p*(1-p))

x = 283.5

cdf_value = stats.norm.cdf(x, loc=mean, scale=std_dev)
print(1-cdf_value)


0.0


Wow! To the accuracy of the computer, the probability that this result was obtained if the true probability of scoring 20-17 was 0.8499% is literally 0. Clearly, this simplified model that assumes independent scores does a terrible job of understanding how scoring actually works in football. But exploring it helped explain some concepts in probability and statistics that we need, and offered insights into how we might create a better model to understand the probability of a scorigami happening. 

# Let's talk about games

Ok, we assumed that each team goes out and scores points independently, and it got us a terrible answer. Clearly, we need to consider the scores as linked together. Let's call the joint probability of a game by $\mathbf{P}_2$, to indicate that we consider this probability as the likelihood that a fixed score of $(a, b)$ (now that we care about the scores happening together, it's natural to assume that $a\geq b$) occurs. Now, we can do exactly what we did before, and construct an empirical distribution of the probability that a given score $(a, b)$ happens. Only, let's take a look at our data for a minute

In [14]:
df.head()

Score  PtsW  PtsL  Count  Score Probability  Expected Occurrences  \
0  20-17    20    17    284           0.008499            148.109692   
1  27-24    27    24    232           0.005747            100.140308   
2  17-14    17    14    200           0.008109            141.303311   
3  23-20    23    20    199           0.004269             74.384512   
4  24-17    24    17    174           0.008564            149.244089   

   Empirical error  
0       135.890308  
1       131.859692  
2        58.696689  
3       124.615488  
4        24.755911

To find the empirical distribution, we would count the frequency of a score happening and divide by the total number of games. We already have a Count column, so this is just what we need. The empirical probability that one of these scores occurs is simple to find. 

In [16]:
df['Joint Empirical Probability'] = df.apply(lambda row: row['Count']/num_games, axis=1)
df.head()

Score  PtsW  PtsL  Count  Score Probability  Expected Occurrences  \
0  20-17    20    17    284           0.008499            148.109692   
1  27-24    27    24    232           0.005747            100.140308   
2  17-14    17    14    200           0.008109            141.303311   
3  23-20    23    20    199           0.004269             74.384512   
4  24-17    24    17    174           0.008564            149.244089   

   Empirical error  Joint Empirical Probability  
0       135.890308                     0.016297  
1       131.859692                     0.013313  
2        58.696689                     0.011477  
3       124.615488                     0.011420  
4        24.755911                     0.009985

Now when we want to see how many games we would expect to have occured with a certain score, we multiply the probability by the total number of games that have occurred: 

In [17]:
df['Joint Expected Occurrences'] = df.apply(lambda row: row['Joint Empirical Probability']*num_games, axis = 1)
df['Joint Empirical Error'] = df.apply(lambda row: abs(row['Joint Expected Occurrences'] - row['Count']), axis = 1)
df.head()

Score  PtsW  PtsL  Count  Score Probability  Expected Occurrences  \
0  20-17    20    17    284           0.008499            148.109692   
1  27-24    27    24    232           0.005747            100.140308   
2  17-14    17    14    200           0.008109            141.303311   
3  23-20    23    20    199           0.004269             74.384512   
4  24-17    24    17    174           0.008564            149.244089   

   Empirical error  Joint Empirical Probability  Joint Expected Occurrences  \
0       135.890308                     0.016297                       284.0   
1       131.859692                     0.013313                       232.0   
2        58.696689                     0.011477                       200.0   
3       124.615488                     0.011420                       199.0   
4        24.755911                     0.009985                       174.0   

   Joint Empirical Error  
0                    0.0  
1                    0.0  
2                    0.0  
3                    0.0  
4                    0.0

Aha, so we've gotten it perfectly right! There is no error between what we expect and what we have observed...except wait a minute. All we did was divide by the number of games to find the probability, and then multiply by the number of games to find how many we expect to have seen. Of course this yields the original number. 

So what's the problem? After all, it feels pretty good to say that we can look to historical data and understand how likely each score is. 

But the problem occurs when we want to know the probability of a scorigami. If you have never seen the result before, an empirical probability distribution will only ever be able to assign the probability of this event occuring as 0, even if you know that it's still possible to occur. 

Although it was flawed, the first scenario got around this. Even if a score, say, 16-5 has never been recorded, teams have scored both 16 and 5 points before. The model where team scores are independent would assign a positive probability to this outcome--albeit a small one. However, it too was limited--for instance, no team has ever score 67 points in a game, although it is certainly possible. 

# Let's bring out Bayes

How do you assess the probability of events that have never occurred? For instance, what is the probability that your car gets stolen tonight? For most of us, we have never had a car get stolen. This makes it impossible to use our personal information to accurately assess this. But insurance companies are able to assess this--they have data on cars getting stolen along with various risk factors that inform their $\textit{belief}$ about rare events occurring. The point is that when data is scarce, we need to incorporate whatever additional information we have to understand something. Then, if new information is presented, we can update our beliefs accordingly. This philosophy is the heart of what is known as Bayesian statistics. 

The information that we belief about how likely events are is called a $\textit{prior}$ distribution. It is subjective--there is no one way to compute a prior. I will present a method of creating a prior distribution for football scoring that incorporates both the rules of the game and information about how likely various scoring events are. 

# Constructing the prior

To construct a prior distribution on scoring, let us create a basic model of scoring. First, note that at any one time in football, a given team has possession of the football. In this chart, the team that has possession is on the left, and the team that does not have possession is on the right. To define possession, we will say that a team has possession if they have possession when the play begins (so if a punt return, pick 6, fumble return, kick return, etc., happens, the scoring counts for the defensive team)

With Ball || Without Ball

(0, 0)- no score (this represents a possession in which neither team scores, and ends in a turnover/punt/end of half/game)

(3, 0)- the team with the ball scores a field goal

(6, 0)- the team with the ball scores a touchdown and misses the extra point/conversion

(7, 0)- the team with the ball scores a touchdown and gets an extra point

(8, 0)- the team with the ball scores a touchdown and gets a two-point conversion

(0, 2)- the team on defense gets a safety

(0, 6)- the team on defense scores a defensive touchdown (return of some sort) and misses the extra point/conversion

(0, 7)- the team on defense scores a defensive touchdown and gets an extra point 

(0, 8)- the team on defense scores a defensive touchdown and gets a two-point conversion

(6, 2)- the team on offense scores a touchdown, and the team on defense returns the conversion for two points

(6, 1)- the team on offense scores a touchdown, and the team on defense scores a defensive safety during the conversion

(2, 6)- the team on defense scores a touchdown, and the team on offense returns the conversion for two points

(1, 6)- the team on defense scores a touchdown, and the team on offense scores a defensive safety during the conversion


The last four options--a touchdown followed by the defensive team scoring, are exceptionally rare. In fact, a defensive safety during a conversion, while possible by the rules, has never happened. Nonetheless, it is possible to envision a scenario in which it could occur: namely, there is a fumble/turnover on the kick or two-point conversion, the defensive player returns it to near the endzone and fumbles. The offensive player attempts to pick up the ball and return it but is tackled for a safety near the endzone. 




To actually construct our prior, what we can do is create a model game in which teams take turns of being in possession of the football, and for every possession, one of the above outcomes occurs. We can empirically estimate the probability of any one of those outcomes occurring by looking at data on the course of football games. Although I will not do so here, one could even model the probability of something occurring, given what is happening in the game (score, time left). Now, the empirical probability of a (6, 1) event happening is 0. To assign this a nonzero probability, let's assume that such a play has a 1% chance of occurring in the next 100 years. There are around 20 drives in an average game, and 272 games in a season, so there are around 20*272*100= 544,000 drives in the next 100 years. Since it has a 1% chance of occurring, we'll set our probability to 0.01/544000 $\approx 2*10^{-8}$. 